### PR TITLE
Implementation limits and Type 4 functions #795

### PR DESCRIPTION
**Pull request by ArtemKiriutin**
Added function parser for parsing PDFunction elements.

### CODE REVIEW

File: src/main/java/org/verapdf/pd/function/PDFunction.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import java.util.logging.Logger;
+
+public class PDFunction extends PDObject {
+
+    private static final Logger LOGGER = Logger.getLogger(PDFunction.class.getCanonicalName());
+
+    private final COSStream cosFunctionStream;
+    private final Long functionType;
+
+    public PDFunction(COSStream cosFunctionStream) {
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/function/PDFunction.java]: rewrite

File: src/main/java/org/verapdf/pd/function/FunctionKeywords.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
+package org.verapdf.pd.function;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public class FunctionKeywords {
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/function/FunctionKeywords.java]: move constants into the parser class

File: src/main/java/org/verapdf/pd/colors/PDSeparation.java [hunk 3]
```diff
@@ -45,15 +52,24 @@ public PDColorSpace getAlternate() {
         return super.getBaseColorSpace();
     }
 
+    public COSObject getTintTransform() {
+        return getObject().at(3);
+    }
+
+    public List<PDFunction> getCustomTintTransform() {
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/colors/PDSeparation.java]: return single function

File: src/main/java/org/verapdf/pd/colors/PDDeviceN.java [hunk 4]
```diff
@@ -62,6 +64,15 @@ public COSObject getTintTransform() {
         return getObject().at(3);
     }
 
+    public List<PDFunction> getCustomTintTransform() {
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/colors/PDDeviceN.java]: return single function, rename method

File: src/main/java/org/verapdf/pd/colors/PDDeviceN.java [hunk 5]
```diff
@@ -62,6 +64,15 @@ public COSObject getTintTransform() {
...
+        COSObject cosFunction = getTintTransform();
+        List<PDFunction> resultList = new ArrayList<>();
+        if (cosFunction != null && cosFunction.getType().equals(COSObjType.COS_STREAM)) {
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/colors/PDDeviceN.java]: enum comparing doesn't require equals

File: src/main/java/org/verapdf/pd/PDHalftone.java [hunk 6]
```diff
@@ -52,4 +58,18 @@ public String getHalftoneName() {
         }
         return base.getStringKey(ASAtom.HALFTONE_NAME);
     }
+
+    public List<PDFunction> getTransferFunction() {
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/PDHalftone.java]: return single object

File: src/main/java/org/verapdf/pd/PDHalftone.java [hunk 7]
```diff
@@ -52,4 +58,18 @@ public String getHalftoneName() {
...
+        COSObject transferFunction = getObject().getKey(ASAtom.getASAtom("transferFunction"));
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/PDHalftone.java]: create a new asatom object in ASAtom class

File: src/main/java/org/verapdf/pd/PDExtGState.java [hunk 8]
```diff
@@ -125,4 +128,34 @@ public PDFont getFont() {
...
+        COSObject trFunction = getKey(ASAtom.TR);
+        COSObject tr2Function = getKey(ASAtom.TR2);
+
+        resultList.addAll(getTRByObj(trFunction));
+        resultList.addAll(getTRByObj(tr2Function));
+
+        return resultList.isEmpty() ? Collections.emptyList() : Collections.unmodifiableList(resultList);
+    }
+
+    private List<PDFunction> getTRByObj(COSObject obj) {
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/PDExtGState.java]: rename, refactor method

File: src/main/java/org/verapdf/parser/Token.java [hunk 9]
```diff
@@ -96,7 +96,7 @@ public void clearValue() {
 		KW_F,
 		KW_XREF,
 		KW_STARTXREF,
-		KW_TRAILER
+		KW_TRAILER,
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/parser/Token.java]: revert

File: src/main/java/org/verapdf/pd/function/PDFunction.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    private static final Logger LOGGER = Logger.getLogger(PDFunction.class.getCanonicalName());
+
+    private final Long functionType;
+
+    protected PDFunction(COSObject obj) {
+       super(obj);
+       this.functionType = obj.getIntegerKey(ASAtom.FUNCTION_TYPE);
+   }
+
+   public static PDFunction createFunction(COSObject obj) {
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/function/PDFunction.java]: the obj should always be a dictionary based type

File: src/main/java/org/verapdf/pd/function/PDFunction.java [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+import java.util.Collections;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class PDFunction extends PDObject {
+
+    private static final Logger LOGGER = Logger.getLogger(PDFunction.class.getCanonicalName());
+
+    private final Long functionType;
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/function/PDFunction.java]: remove this field

Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/function/PDFunction.java]: add FunctionParser field

File: src/main/java/org/verapdf/pd/function/PDType3Function.java [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+		}
+		List<COSObject> resultList = new ArrayList<>();
+
+		for (COSObject cosObj : functions.getValues()){
+			resultList.addAll(super.getInnerNumbers(cosObj));
+		}
+		return Collections.unmodifiableList(resultList);
+	}
+
+	@Override
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/function/PDType3Function.java]: remove

File: src/main/java/org/verapdf/pd/function/PDType3Function.java [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+import org.verapdf.cos.COSObjType;
+import org.verapdf.cos.COSObject;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+public class PDType3Function extends PDFunction {
+
+	private static final int TYPE3_FUNCTION = 3;
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/function/PDType3Function.java]: remove

File: src/main/java/org/verapdf/pd/function/PDType3Function.java [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+
+	protected PDType3Function(COSObject obj) {
+		super(obj);
+	}
+
+	@Override
+	public List<COSObject> getInnerNumbers(COSObject functions) {
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/function/PDType3Function.java]: remove

File: src/main/java/org/verapdf/pd/function/PDFunction.java [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+            case COS_ARRAY:
+                return new PDType3Function(obj);
+            case COS_STREAM:
+                return new PDFunction(obj);
+            default:
+                return null;
+        }
+   }
+
+    public List<COSObject> getInnerNumbers(COSObject function) {
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/function/PDFunction.java]: refactor, keep in mind that the function object can be not a stream

File: src/main/java/org/verapdf/pd/function/PDType3Function.java [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+			return Collections.emptyList();
+		}
+		List<COSObject> resultList = new ArrayList<>();
+
+		for (COSObject cosObj : functions.getValues()){
+			resultList.addAll(super.getInnerNumbers(cosObj));
+		}
+		return Collections.unmodifiableList(resultList);
+	}
+
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/function/PDType3Function.java]: add `getFunctions` method

File: src/main/java/org/verapdf/pd/colors/PDSeparation.java [hunk 17]
```diff
@@ -45,15 +52,24 @@ public PDColorSpace getAlternate() {
         return super.getBaseColorSpace();
     }
 
+    public COSObject getTintTransform() {
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/colors/PDSeparation.java]: rename to `getCosTintTransform`

File: src/main/java/org/verapdf/pd/colors/PDSeparation.java [hunk 18]
```diff
@@ -45,15 +52,24 @@ public PDColorSpace getAlternate() {
...
+        return getObject().at(3);
+    }
+
+    public PDFunction getCustomTintTransform() {
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/colors/PDSeparation.java]: rename to `getTintTransform`

File: src/main/java/org/verapdf/pd/PDExtGState.java [hunk 19]
```diff
@@ -125,4 +131,39 @@ public PDFont getFont() {
         }
         return null;
     }
+
+    public List<PDFunction> getTRFunctions() {
+        return Collections.unmodifiableList(getFunctions(ASAtom.TR));
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/PDExtGState.java]: simple return getFunctions as unmodifialbe list created there

File: src/main/java/org/verapdf/pd/PDExtGState.java [hunk 20]
```diff
@@ -125,4 +131,39 @@ public PDFont getFont() {
...
+    }
+
+    public List<PDFunction> getTR2Functions() {
+        return Collections.unmodifiableList(getFunctions(ASAtom.TR2));
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/PDExtGState.java]: simple return getFunctions as unmodifialbe list created there

File: src/main/java/org/verapdf/pd/PDExtGState.java [hunk 21]
```diff
@@ -125,4 +131,39 @@ public PDFont getFont() {
...
+    }
+
+    private List<PDFunction> getFunctions(ASAtom atom) {
+        COSObject functions = getKey(atom);
+        if (functions.getType() == COSObjType.COS_STREAM) {
+            PDFunction function = PDFunction.createFunction(functions);
+            if (function == null) {
+                return Collections.emptyList();
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/PDExtGState.java]: `if (function != null) {return ...}`, else is not necesasry

File: src/main/java/org/verapdf/pd/PDExtGState.java [hunk 22]
```diff
@@ -125,4 +131,39 @@ public PDFont getFont() {
...
+
+    public List<PDFunction> getTRFunctions() {
+        return Collections.unmodifiableList(getFunctions(ASAtom.TR));
+    }
+
+    public List<PDFunction> getTR2Functions() {
+        return Collections.unmodifiableList(getFunctions(ASAtom.TR2));
+    }
+
+    private List<PDFunction> getFunctions(ASAtom atom) {
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/PDExtGState.java]: minor clean code refactoring

File: src/main/java/org/verapdf/pd/function/PDFunction.java [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+    protected PDFunction(COSObject obj) {
+       super(obj);
+   }
+
+   public static PDFunction createFunction(COSObject obj) {
+        if (obj == null || !obj.getType().isDictionaryBased()) {
+            return null;
+        }
+
+        if (obj.getIntegerKey(ASAtom.FUNCTION_TYPE) == null) {
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/function/PDFunction.java]: extract Integer into a variable, use that variable.intValue() in switch

File: src/main/java/org/verapdf/pd/function/PDFunction.java [hunk 24]
```diff
@@ -0,0 +1,69 @@
...
+            case 3:
+                return new PDType3Function(obj);
+            default:
+                return new PDFunction(obj);
+        }
+   }
+
+    public List<COSObject> getInnerNumbers() {
+        COSObject function = this.getObject();
+        if (function == null || function.getType() != COSObjType.COS_STREAM) {
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/function/PDFunction.java]: function cannot be null

File: src/main/java/org/verapdf/pd/function/PDFunction.java [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+        switch (functionType) {
+            case 3:
+                return new PDType3Function(obj);
+            default:
+                return new PDFunction(obj);
+        }
+   }
+
+    public List<COSObject> getInnerNumbers() {
+        COSObject function = this.getObject();
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/function/PDFunction.java]: rename function

File: src/main/java/org/verapdf/pd/function/PDType3Function.java [hunk 26]
```diff
@@ -0,0 +1,31 @@
...
+public class PDType3Function extends PDFunction {
+
+	protected PDType3Function(COSObject obj) {
+		super(obj);
+	}
+
+	public List<PDFunction> getFunctions() {
+
+		COSObject functions = getObject();
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/function/PDType3Function.java]: incorrect logic (getKey(ASAtom.Function)

File: src/main/java/org/verapdf/parser/FunctionParser.java [hunk 27]
```diff
@@ -0,0 +1,111 @@
...
+                }
+                break;
+            case TT_INTEGER:
+                functionObjects.add(COSInteger.construct(this.getToken().integer));
+                break;
+            case TT_REAL:
+                functionObjects.add(COSReal.construct(this.getToken().real));
+                break;
+            default:
+                LOGGER.warning("Invalid object type in Function");
```
[the remark this answers is not in the document]
Comment (ArtemKiriutin) [reply src/main/java/org/verapdf/parser/FunctionParser.java]: Checked. There cannot be other token types.

File: src/main/java/org/verapdf/pd/PDExtGState.java [hunk 28]
```diff
@@ -125,4 +131,35 @@ public PDFont getFont() {
...
+        COSObject functions = getKey(atom);
+        if (functions.getType() == COSObjType.COS_STREAM) {
+            PDFunction function = PDFunction.createFunction(functions);
+            if (function != null) {
+                return Collections.singletonList(function);
+            }
+        } else if (functions.getType() == COSObjType.COS_ARRAY) {
+            List<PDFunction> resultList = new ArrayList<>();
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/PDExtGState.java]: move before for

File: src/main/java/org/verapdf/pd/function/PDFunction.java [hunk 29]
```diff
@@ -0,0 +1,66 @@
...
+
+   public static PDFunction createFunction(COSObject obj) {
+        if (obj == null || !obj.getType().isDictionaryBased()) {
+            return null;
+        }
+
+       Long functionType = obj.getIntegerKey(ASAtom.FUNCTION_TYPE);
+
+        if (functionType == null) {
+            return null;
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/function/PDFunction.java]: add log (warning) message, create new PDFunction(obj)

File: src/main/java/org/verapdf/pd/function/PDFunction.java [hunk 30]
```diff
@@ -0,0 +1,66 @@
...
+        COSObject obj = this.getObject();
+        if (obj.getType() != COSObjType.COS_STREAM) {
+            return Collections.emptyList();
+        }
+        if (this.parser == null) {
+            try(ASInputStream functionStream = obj.getData()) {
+                this.parser = new FunctionParser(functionStream);
+                this.parser.parse();
+            } catch (IOException e) {
+                LOGGER.log(Level.WARNING,"Can not parse Type4 function", e);
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/function/PDFunction.java]: `"Can not parse function"`

File: src/main/java/org/verapdf/pd/function/PDFunction.java [hunk 31]
```diff
@@ -0,0 +1,66 @@
...
+        }
+   }
+
+    public List<COSObject> getInnerNumbers() {
+        COSObject obj = this.getObject();
+        if (obj.getType() != COSObjType.COS_STREAM) {
+            return Collections.emptyList();
+        }
+        if (this.parser == null) {
+            try(ASInputStream functionStream = obj.getData()) {
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/function/PDFunction.java]: DECODE

File: src/main/java/org/verapdf/pd/function/PDType3Function.java [hunk 32]
```diff
@@ -0,0 +1,37 @@
...
+
+	protected PDType3Function(COSObject obj) {
+		super(obj);
+	}
+
+	public List<PDFunction> getFunctions() {
+
+		COSObject obj = getKey(ASAtom.FUNCTIONS);
+		if(obj == null || obj.getType() != COSObjType.COS_ARRAY) {
+			LOGGER.warning("Value of Functions key is null or not an array");
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/function/PDType3Function.java]: `"Invalid Functions value in Type 3 Function dictionary"`

File: src/main/java/org/verapdf/parser/FunctionParser.java [hunk 33]
```diff
@@ -0,0 +1,130 @@
...
+        tempSet.add("dup");
+        tempSet.add("index");
+        tempSet.add("roll");
+        tempSet.add("{");
+        tempSet.add("}");
+
+        FUNCTION_KEYWORDS = Collections.unmodifiableSet(tempSet);
+    }
+
+    private List<COSObject> functionNumbers = new ArrayList<>();
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/parser/FunctionParser.java]: isn't it operators?

File: src/main/java/org/verapdf/pd/function/PDFunction.java [hunk 34]
```diff
@@ -0,0 +1,88 @@
...
+   }
+
+    public List<COSObject> getOperators() {
+        COSObject obj = this.getObject();
+        if (obj.getType() != COSObjType.COS_STREAM) {
+            return Collections.emptyList();
+        }
+        if (this.parser == null) {
+            try (ASInputStream functionStream = obj.getData(COSStream.FilterFlags.DECODE)) {
+                this.parser = new FunctionParser(functionStream);
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/function/PDFunction.java]: Add this.parser.closeStream() in finally (if it is not null)

File: src/main/java/org/verapdf/pd/function/PDFunction.java [hunk 35]
```diff
@@ -0,0 +1,96 @@
...
+                LOGGER.log(Level.WARNING,"Can not parse function", e);
+                return Collections.emptyList();
+            } finally {
+                if (this.parser != null) {
+                    try {
+                        this.parser.closeInputStream();
+                    } catch (IOException e) {
+                        LOGGER.log(Level.WARNING, "Can not close input stream", e);
+                    }
+                }
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/function/PDFunction.java]: `else {return Collections.emptyList();}` with comment

File: src/main/java/org/verapdf/pd/function/PDFunction.java [hunk 36]
```diff
@@ -0,0 +1,96 @@
...
+        }
+   }
+
+    public List<COSObject> getOperators() {
+        COSObject obj = this.getObject();
+        if (obj.getType() != COSObjType.COS_STREAM) {
+            return Collections.emptyList();
+        }
+        if (this.parser == null) {
+            try (ASInputStream functionStream = obj.getData(COSStream.FilterFlags.DECODE)) {
```
Comment (BezrukovM) [inline src/main/java/org/verapdf/pd/function/PDFunction.java]: extract parsing into private method. catch exception here

**Pull request merged**